src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+
+            return ConvertToSemanticTokenEdits(newGroupedSemanticTokens, edits);
+        }
+
+        private static SemanticTokensEdit[] ConvertToSemanticTokenEdits(SemanticToken[] newGroupedSemanticTokens, IEnumerable<SequenceEdit> edits)
+        {
+            // Our goal is to minimize the number of edits we return to LSP. It's possible an index
+            // may have both an insertion and deletion, in which case we can combine the two into a
+            // single update. We use the dictionary below to keep track of whether an index contains
+            // an insertion, deletion, or both.
+            var indexToEditKinds = new Dictionary<int, SemanticTokenEditKind>();
+
+            foreach (var edit in edits)
+            {
+                // We only care about EditKind.Insert and EditKind.Delete, since they encompass all
+                // changes to the document. All other EditKinds are ignored.
+                switch (edit.Kind)
+                {
+                    case EditKind.Insert:
+                        indexToEditKinds.TryGetValue(edit.NewIndex, out var editKindWithoutInsert);
+                        indexToEditKinds[edit.NewIndex] = editKindWithoutInsert |= SemanticTokenEditKind.Insert;
+                        break;
+                    case EditKind.Delete:
+                        indexToEditKinds.TryGetValue(edit.OldIndex, out var editKindWithoutDelete);
+                        indexToEditKinds[edit.OldIndex] = editKindWithoutDelete |= SemanticTokenEditKind.Delete;
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            return CombineEditsIfPossible(newGroupedSemanticTokens, indexToEditKinds);
+        }
+
+        private static SemanticTokensEdit[] CombineEditsIfPossible(
+            SemanticToken[] newGroupedSemanticTokens,
+            Dictionary<int, SemanticTokenEditKind> indexToEditKinds)
+        {
+            // This method combines the edits into the minimal possible edits.
+            // For example, if an index contains both an insertion and deletion, we can combine the two
+            // edits into one.
+            // We can also combine edits if we have consecutive edits of certain types:
+            // Delete->Delete, Insert->Insert, Update->Update, Update->Insert, and Update->Delete.
+            // Note for the Update->Insert and Update->Delete cases, any further edits we combine can
+            // only be an Insert or Delete, respectively.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
```

Comment:
Indices is plural of Index

Reply:
so these keys can be in a random order (i.e. dictionaries don't ensure a specific key ordering). is your algorithm correct in the face of that?  should you sort these indices first so we process things in an increasing order?

Reply:
Just a note that indexes and indices are both accepted plural forms of index. Indices, however, is preferred in mathematical or technical fields. https://www.merriam-webster.com/dictionary/index

Reply:
https://media2.giphy.com/media/3ohzdQChVBWTRSkILe/giphy.gif